Modules/Keychain/Sources/Secured.swift

```diff
@@ -0,0 +1,173 @@
+            var description: String = .init(format: "An keychain error associated with the key: %@ occurred",key)
+
+            if let errorDescription = self.errorDescription {
+                description.append(String(format: "\nDescription: %@", errorDescription))
+            }
+
+            if let reason = failureReason {
+                description.append(String(format: "\nReason: %@", reason))
+            }
+
+            if let recoverySuggestion = recoverySuggestion {
+                description.append(String(format: "\nRecovery Suggestion: %@", recoverySuggestion))
+            }
+
+            return description
+        }
+    }
+
+    public var wrappedValue: Value? {
+        didSet {
+            do {
+                try storeValueInKeychain(wrappedValue)
+            } catch {
+                wrappedValue = oldValue
+
+                logError(error)
+            }
+        }
+    }
+
+    private var searchQuery: [String: Any] {
+        [
+            kSecClass as String: kSecClassGenericPassword,
+            kSecAttrService as String: key
+        ]
+    }
+
+    private let key: String
+    private let keychain: KeychainProtocol
+
+    /// The default initializer for `Secured`
+    ///
+    /// - Parameter key: The key associated with storing the value inside the Keychain
+    public init(key: String, keychain: KeychainProtocol = Keychain.default) {
+        self.key = key
+        self.keychain = keychain
+
+        do {
+            wrappedValue = try loadValueFromKeychain()
+        } catch {
+            logError(error)
+        }
+    }
+
+    private func logError(_ error: Error) {
+        guard let keychainError = error as? KeychainError else {
+            return NSLog("An unexpected error occurred:\nKey: %@\nError: %@", key, error.localizedDescription)
+        }
+
+        NSLog(keychainError.description(for: key))
+    }
+
+    private func loadValueFromKeychain() throws -> Value {
+        var searchQuery = self.searchQuery
+        searchQuery[kSecReturnAttributes as String] = true
+        searchQuery[kSecReturnData as String] = true
+
+        let keychainResponse = keychain.fetch(searchQuery)
+
+        guard
+            keychainResponse.status != errSecItemNotFound,
+            keychainResponse.status == errSecSuccess,
+            let item = keychainResponse.queryResult as? [String: Any],
+            let data = item[kSecValueData as String] as? Data
+        else { throw KeychainError.itemNotFound }
+
+        do {
+            let value = try JSONDecoder().decode(Value.self, from: data)
+            return value
+        } catch {
+            throw KeychainError.decodingError(error: error)
+        }
+    }
+
+    private func storeValueInKeychain(_ value: Value?) throws {
+        guard let value = value else { return try deleteFromKeychain() }
+
+        let encodedData: Data
+        do {
+            encodedData = try JSONEncoder().encode(value)
```

Comment:
Same as above with the JSONDecoder. Should we allow to use a custom encoder here? @JensK611 what do you think?

Reply:
In my opinion it doesn't matter if it is JSON or something else, the point is it has to be serialized as `Data` therefore we can start with the `JSONDecoder`.

Reply:
Okay fine for me as well, just wanted to have it discussed.